<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width,initial-scale=1.0">
        <meta http-equiv="X-UA-Compatible" content="ie=edge">
        <title>取消请求</title>
    </head>
    <body>
        <button onclick="getProduct1()">获取商品列表1</button>
        <button onclick="getProduct2()">获取商品列表2</button>
        <button onclick="cancelReq()">取消请求</button>
    </body>
    <script src="https://cdn.bootcss.com/axios/0.17.1/axios.min.js"></script>
    <script>
        //添加请求拦截器
        axios.interceptors.request.use((config)=>{
            //在准备发请求前，取消未完成的请求
            if(typeof cancel==="function"){
                cancel("强制取消请求");
            }
            //添加一个cancelToken的配置
            config.cancelToken=new axios.CancelToken(function executor(c){//c是用于取消当前请求的函数
                //保存取消函数，用于之后可能需要取消当前请求
                cancel=c;
            });
            return config;
        })

        //添加响应拦截器
        axios.interceptors.response.use(
            (response)=>{
                cancel=null;
                return response;
            },
            (error)=>{
                //请求结束了 就可以干掉cancel了 因为再也用不到了
                //但是进入error的不一定是因为请求结束，也有可能是因为cancel而进入
                //当这种情况进入时候不需要置空cancel，因为还有请求在发送，可能还需要再取消
                //如果因为cancel进入 然后置空了 下一个请求点击时候 就可能无法按预期取消了
                if(axios.isCancel(error)){//取消请求的错误
                    console.log("请求1取消的错误",error.message);//做相应处理
                    //中断promise链接
                    return new Promise(()=>{});
                }else{
                    cancel=null;
                    //将错误向下传递
                    return Promise.reject(error);
                }
                
            })

        let cancel;//用于保存取消请求的函数

        function getProduct1(){
            axios({
                url:"http://localhost:4000/products1"
            }).then(
                response=>{
                    //请求结束了 就可以干掉cancel了 因为再也用不到了
                    console.log("请求1数据成功了",response.data);
                },
                error=>{
                    console.log("请求1失败了",error.message);
                }
            )
        }

        function getProduct2(){
            axios({
                url:"http://localhost:4000/products2"
            }).then(
                response=>{
                    console.log("请求2数据成功了",response.data);
                },
                error=>{
                    console.log("请求2失败了",error.message,error);
                }
            )
        }

        function cancelReq(){
            console.log("cancel request");
            //执行取消请求的函数
            if(typeof cancel==='function'){
                cancel("强制取消请求");
            }else{
                console.log("没有可取消的请求");
            }
            
        }
    </script>
</html>